File: bbq-core/src/trader/entrust.rs
use chrono::NaiveDateTime;
use serde::{Deserialize, Serialize};
use std::fmt::{Display, Formatter};
use uuid::Uuid;
use super::signal::Signal;

#[derive(Debug, Clone, serde::Deserialize, serde::Serialize)]
#[serde(rename_all = "snake_case")]
pub enum EntrustStatus {
    Init,
    Commit,
    Deal,
    PartDeal,
    Cancel,
}

impl Default for EntrustStatus {
    fn default() -> Self {
        Self::Init
    }
}

impl Display for EntrustStatus {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        let s = match self {
            EntrustStatus::Init => "初始化",
            EntrustStatus::Commit => "已提交",
            EntrustStatus::Deal => "已成交",
            EntrustStatus::PartDeal => "部分成交",
            EntrustStatus::Cancel => "已撤销",
        };
        write!(f, "{}", s)
    }
}

#[derive(Debug, Clone, serde::Deserialize, serde::Serialize)]
#[serde(rename_all = "snake_case")]
pub enum EntrustType {
    Buy,
    Sell,
    Cancel,
}

impl Default for EntrustType {
    fn default() -> Self {
        Self::Cancel
    }
}

impl Display for EntrustType {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        let s = match self {
            EntrustType::Sell => "卖出",
            EntrustType::Buy => "买入",
            EntrustType::Cancel => "撤销",
        };
        write!(f, "{}", s)
    }
}

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
#[serde(deny_unknown_fields, default, rename_all = "snake_case")]
pub struct Entrust {
    pub entrust_id: String,
    pub name: String,
    pub code: String,
    pub time: Option<NaiveDateTime>,

    pub entrust_type: EntrustType,
    pub status: EntrustStatus,

    pub price: f64,
    pub volume: u32,

    pub volume_deal: u32,
    pub volume_cancel: u32,

    pub desc: String,

    pub broker_entrust_id: Option<String>,
}

impl Entrust {
    pub fn new_from_signal(signal: &Signal) -> Self {
        Self {
            entrust_id: Uuid::new_v4().to_simple().to_string(),
            name: signal.name.clone(),
            code: signal.code.clone(),
            time: signal.time.clone(),
            entrust_type: match signal.signal {
                super::signal::SignalType::Sell => EntrustType::Sell,
                super::signal::SignalType::Buy => EntrustType::Buy,
                super::signal::SignalType::Cancel => EntrustType::Cancel,
            },
            status: EntrustStatus::Init,
            price: signal.price,
            volume: signal.volume,
            volume_deal: 0,
            volume_cancel: 0,
            desc: "".to_string(),
            broker_entrust_id: None,
        }
    }
}

File: bbq-strategy-py/qstrategy/entrust.py
from datetime import datetime


class Entrust:
    st_init, st_commit, st_deal, st_part_deal, st_cancel = 'init', 'commit', 'deal', 'part_deal', 'cancel'
    typ_buy, typ_sell, typ_cancel = 'buy', 'sell', 'cancel'
    def __init__(self, js):
        self.entrust_id = js['entrust_id']
        self.name = js['name']
        self.code = js['code']
        self.volume_deal = js['volume_deal']
        self.volume_cancel = js['volume_cancel']
        self.volume = js['volume']
        self.price = js['price']
        self.status = js['status']
        self.entrust_type = js['entrust_type']
        self.desc = js['desc']
        self.broker_entrust_id = js['broker_entrust_id'] if 'broker_entrust_id' in js else None

        self.time = None
        if js['time'] is not None and isinstance(js['time'], str):
            s = js['time'][:-7]
            self.time = js['time']
            self.time = datetime.strptime(s, '%Y-%m-%dT%H:%M:%S')

    def to_dict(self):
        d = {}
        d['entrust_id'] = self.entrust_id
        d['name'] = self.name
        d['code'] = self.code

        d['volume_deal'] = self.volume_deal
        d['volume_cancel'] = self.volume_cancel
        d['volume'] = self.volume
        d['price'] = self.price
        d['status'] = self.status
        d['entrust_type'] = self.entrust_type
        d['desc'] = self.desc
        d['broker_entrust_id'] = self.broker_entrust_id

        d['time'] = None if self.time is None else self.time.strftime(
            "%Y-%m-%dT%H:%M:%S.%f")

        return d
